crates/agent/src/publications/specs.rs

```diff
@@ -546,3 +546,174 @@ fn split_tag(image_full: &str) -> (String, String) {
         (image, String::new())
     }
 }
+
+#[cfg(test)]
+mod test {
+
+    use crate::publications::JobStatus;
+
+    use super::super::PublishHandler;
+    use agent_sql::Id;
+    use reqwest::Url;
+    use sqlx::{Connection, Postgres, Transaction};
+
+    // Squelch warnings about struct fields never being read.
+    // They actually are read by insta when snapshotting.
+    #[allow(dead_code)]
+    #[derive(Debug)]
+    struct ScenarioResult {
+        draft_id: Id,
+        status: JobStatus,
+        errors: Vec<String>,
+    }
+
+    const FIXED_DATABASE_URL: &str = "postgresql://postgres:postgres@localhost:5432/postgres";
+
+    const CLEANUP: &str = r#"
+    with specs_delete as (
+        delete from live_specs
+    ),
+    drafts_delete as (
+        delete from drafts
+    ),
+    draft_specs_delete as (
+        delete from draft_specs
+    ),
+    publications_delete as (
+        delete from publications
+    ),
+    role_grants_delete as (
+        delete from role_grants
+    ),
+    user_grants_delete as (
+        delete from user_grants
+    ),
+    tags_delete as (
+        delete from connector_tags
+    ),
+    connectors_delete as (
+        delete from connectors
+    )
+    select 1;
+    "#;
+
+    async fn scenario_snapshot<'c>(
+        scenario: &str,
+        mut txn: Transaction<'c, Postgres>,
+    ) -> Vec<ScenarioResult> {
+        sqlx::query(CLEANUP).execute(&mut txn).await.unwrap();
+        sqlx::query(scenario).execute(&mut txn).await.unwrap();
+
+        let bs_url: Url = "http://example.com".parse().unwrap();
+
+        let (logs_tx, mut logs_rx) = tokio::sync::mpsc::channel(8192);
+
+        // Just in case anything gets through
+        logs_rx.close();
+
+        let mut handler = PublishHandler::new("", &bs_url, &bs_url, "", &bs_url, &logs_tx);
+
+        let mut results: Vec<ScenarioResult> = vec![];
+
+        while let Some(row) = agent_sql::publications::dequeue(&mut txn).await.unwrap() {
+            let mut sub_tx = txn.begin().await.unwrap();
```

Comment:
The thought was that it would leave a clean slate for every publication, but I realized that it's all happening within one test run, so this is the wrong thing to do... also I was committing the transaction instead of rolling it back, defeating the purpose anyway :P